drivers/base/nmidebug/nmidebug.c

```diff
@@ -13,49 +12,31 @@
 
 /* FUNCTIONS ******************************************************************/
 
-PCHAR NmiBegin = "NMI4NMI@";
-
-FORCEINLINE
-VOID
-NmiClearFlag(VOID)
-{
-    ((PCHAR)&KiBugCheckData[4])[0] -= (NmiBegin[3] | NmiBegin[7]);
-    ((PCHAR)&KiBugCheckData[4])[3] |= 1;
-#ifdef _M_IX86
-#if defined(_MSC_VER) && !defined(__clang__)
-    __asm
-    {
-        rcr KiBugCheckData[4], 8
-    }
-#else
-    __asm__("rcrl %b[shift], %k[retval]" : [retval] "=rm" (KiBugCheckData[4]) : "[retval]" (KiBugCheckData[4]), [shift] "Nc" (8));
-#endif
-#endif
-}
-
 BOOLEAN
 NTAPI
-NmiDbgCallback(IN PVOID Context,
-               IN BOOLEAN Handled)
+NmiDbgCallback(
+    _In_opt_ PVOID Context,
+    _In_ BOOLEAN Handled)
 {
-    /* Clear the NMI flag */
-    NmiClearFlag();
+    /* Handle the NMI safely if a debugger is connected */
+    if (!KD_DEBUGGER_NOT_PRESENT && KD_DEBUGGER_ENABLED)
+    {
+        DbgPrint("NMI Callback entered!\n");
 
-    /* Get NMI status signature */
-    __indwordstring(0x80, (PULONG)NmiBegin, 1);
-    ((void(*)())&KiBugCheckData[4])();
+        DbgBreakPoint();
```

Comment:
In https://community.osr.com/discussion/comment/286830/#Comment_286830 I see
> Since an NMI can interrupt anything (and there is almost no code that can safely be called from an NMI callback), synchronizing with any code that is called to handle an NMI without risk of deadlock etc. is difficult and cumbersome, not to mention that NMI code typically cannot be debugged in typical fashion, as the debugger itself relies on NMIs to function (at least on AMD64).

I cannot judge of the exactness of this claim but, if this is true, how is this breakpoint breaking in windbg going to behave?

Reply:
That's interesting... On ROS this patch makes the debugger to break at every NMI invocation.
I have tested also on checked windows 2003 x86 but it doesn't break into the debugger and only prints the message "Port lock was not acquired!".

Reply:
Removed call to DbgBreakPoint.